File: kattis/walrus_weights.cpp
#include <iostream>
#include <cstring>

using namespace std;

int n;
int c[1005];
bool dp[2005][1005];

int main () {
  memset(dp, false, 2005 * 1005);

  cin >> n;
  for (int i = 0; i < n; i++)
    cin >> c[i];

  for (int i = 0; i < n; i++)
    dp[0][i] = true;

  dp[c[0]][0] = true;

  for (int i = 1; i < 2005; i++) {
    for (int j = 1; j < n; j++) {
      if (i - c[j] >= 0 && dp[i - c[j]][j-1])
        dp[i][j] = true;
      dp[i][j] |= dp[i][j-1];
    }
  }

  int r = 0;

  for (int i = 0; i < 2005; i++) {
    if (dp[i][n-1] && abs(1000 - i) <= abs(1000 - r))
      r = i;
  }

  cout << r << endl;

  return 0;
}

File: kattis/walrus_weights.py
n = int(input())
c = [int(input()) for _ in range(n)]

dp = [[False for _ in range(n)] for _ in range(2005)]

for i in range(n):
    dp[0][i] = True

dp[c[0]][0] = True

for i in range(1, 2005):
    for j in range(1, n):
        if i - c[j] >= 0 and dp[i - c[j]][j-1]:
            dp[i][j] = True
        dp[i][j] |= dp[i][j-1]

r = 0

for i in range(2005):
    if dp[i][n-1] and abs(1000 - i) <= abs(1000 - r):
        r = i

print(r)
